### PR TITLE
First attend to implement compute resource view and entity

### PR DESCRIPTION
**Pull request by schlupov**
Test example
```
from nailgun import entities

from robottelo.datafactory import gen_string, valid_data_list
from robottelo.decorators import parametrize

@parametrize('name', valid_data_list())
def test_positive_create_docker(session, name):
    """Create new CR, provider is Docker"""
    with session:
        session.computeresource.create({
        'name': name,
        'provider':'Docker',
        'url':'https://docker.example.com:4243'
        })
        assert session.computeresource.search(name) == name

@parametrize('name', valid_data_list())
def test_positive_create_libvirt(session, name):
    """Create new CR, provider is Libvirt"""
    with session:
        session.computeresource.create({
        'name': name,
        'provider':'Libvirt',
        'url':'qemu://host.example.com/system',
        'display_type':'VNC',
        })
        assert session.computeresource.search(name) == name
```

### CODE REVIEW

**Comment by schlupov** [comment]
This is not a final version I just want you to give me a feedback

**Comment by abalakh** [comment]
I'm curios what the best approach here as well.
@mfalesni, could you please provide your suggestions?
We have following UI form:
screen shot 2018-03-27 at 11 52 58 am
Which expands depending on selected provider, eg:
screen shot 2018-03-27 at 11 53 21 am
or
screen shot 2018-03-27 at 11 53 10 am
I can see a lot of possible options, like nested views, included views, custom fill, etc, but what's the optimal way to handle such forms?

**Comment by mfalesni** [comment]
@abalakh This looks like a simple case. Just put all kinds of fields in the view. Unless there is a clash, like Foo input vs Foo dropdown, which would necessitate use of switchable views/widgets, this is fine. Just make sure the order of definition is right.

**Comment by jhutar** [comment]
@schlupov , so, looks like you are going a right direction for now.

**Comment by schlupov** [comment]
I'm so sorry I didn't answer earlier. Thank you for your help with that.

**Comment by schlupov** [comment]
@abalakh Please can you take a look at my code if it's fine? I don't know what I should do with my tests, they contain my internal hostnames and passwords. Also can you tell me if I should add some more properties into views/entities?

**Comment by abalakh** [comment]
@schlupov your tests shouldn't have hardcoded internal url or passwords. Please fill those values into corresponding sections of your robottelo.properties file and reuse values from properties file in your tests. It's easier than it sounds and there's bunch of examples, e.g. https://github.com/SatelliteQE/robottelo/blob/master/tests/foreman/ui/test_computeresource_ec2.py#L47 or https://github.com/SatelliteQE/robottelo/blob/master/tests/foreman/ui/test_computeresource_vmware.py#L48

Regarding amount of widgets in views - no point in trying to cover everything, that's too time consuming and you probably won't need every single UI widget in your tests. Just cover all the controls and inputs required for tests and that's it, if we need more widgets later - we'll add them altogether with tests targeting them.

**Comment by schlupov** [comment]
Thank you for the quick response. I'll take a look at that lines you mentioned above.

File: airgun/entities/computeresource.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        view = self.navigate_to(self, 'Edit', entity_name=name)
+        view.fill(values)
+        view.submit.click()
+
+    def read(self, entity_name):
+        view = self.navigate_to(self, 'Edit', entity_name=entity_name)
+        return view.read()
+
+    def delete(self, entity_name):
+        view = self.navigate_to(self, 'Delete', entity_name=entity_name)
```
Comment (oshtaier) [inline airgun/entities/computeresource.py]: I wouldn't say that delete is a separate navigation point, so in my opinion it should be like this:
```
view = self.navigate_to(self, 'All')
view.search(value)
view.action_list.fill('Delete')
```

File: airgun/entities/computeresource.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        view.action_list.fill('Delete')
+
+
+@navigator.register(ComputeResourceEntity, 'All')
+class ShowAllActivationComputeResources(NavigateStep):
```
Comment (oshtaier) [inline airgun/entities/computeresource.py]: typo:
`ShowAllComputeResources`

File: airgun/views/computeresource.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    name = TextInput(id='compute_resource_name')
+    provider = FilteredDropdown(id='s2id_compute_resource_provider')
+    description = TextInput(id='compute_resource_description')
+    url = TextInput(id='compute_resource_url')
+    display_type = Select(id='compute_resource_display_type')
+    console_passwords = Checkbox(id='compute_resource_set_console_password')
+    user = TextInput(id='compute_resource_user')
+    password = TextInput(id='compute_resource_password')
+    vcenter = TextInput(id='compute_resource_server')
+    datacenter = Text("//*[@class='caption']//*[text()='Load Datacenters']")
```
[the remark this answers is not in the document]
Comment (schlupov) [reply airgun/views/computeresource.py]: Bad name for variable, should have been something like load_datacenters because it's for the button Load Datacenters but I never needed it anyway (to click on that button).

**Comment by schlupov** [comment]
@oshtaier  Please can you take a look at last changes? I corrected in entities what you didn't like and I don't know if everything is ok now.

File: airgun/views/computeresource.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+from widgetastic.widget import Checkbox, Select, Text, TextInput
+from airgun.views.common import BaseLoggedInView, SearchableViewMixin
+from airgun.widgets import FilteredDropdown, ActionsDropdown
+
+
+class ComputeResourcesView(BaseLoggedInView, SearchableViewMixin):
+    title = Text("//h1[text()='Compute Resources']")
+    new = Text("//a[contains(@href, '/compute_resources/new')]")
+    edit = Text(".//span[contains(@class, 'btn')][a/text()='Edit']")
```
Comment (oshtaier) [inline airgun/views/computeresource.py]: usually we look for that element for edit:
image
If you still like to use that button - it is now a part of `ActionsDropdown` widget

File: airgun/views/computeresource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (oshtaier) [reply airgun/views/computeresource.py]: Hmm, I still think to distinguish each provider type into separate nested views.
```
+    name = TextInput(id='compute_resource_name')
+    provider = FilteredDropdown(id='s2id_compute_resource_provider')
+    description = TextInput(id='compute_resource_description')
```
and
```
    @View.nested
    class docker_provider(View):
        url = TextInput(id='compute_resource_url')
        user = TextInput(id='compute_resource_user')
        password = TextInput(id='compute_resource_password')
        email = TextInput(id='...')
and etc
```

[the remark this answers is not in the document]
Comment (oshtaier) [reply airgun/views/computeresource.py]: @abalakh thoughts?

**Comment by lhellebr** [comment]
Hi, Silva! I've made a pull request to your branch, adding some stuff for testing RHEV. I will add more methods and also tests later. The CR creating might timeout until this tweak is merged: https://github.com/SatelliteQE/airgun/pull/79

File: airgun/entities/computeresource.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        view.submit.click()
+
+    def read(self, entity_name):
+        view = self.navigate_to(self, 'Edit', entity_name=entity_name)
+        return view.read()
+
+    def delete(self, value):
+        view = self.navigate_to(self, 'All')
+        view.search(value)
+        view.action_list.fill('Delete')
```
Comment (abalakh) [inline airgun/entities/computeresource.py]: nitpick: we usually call that widget `actions`, it's shorter.

File: airgun/views/computeresource.py (hunk 4 above)
Comment (abalakh) [reply airgun/views/computeresource.py]: Right, `Edit` should be called via `action_list.fill('Edit')` and i'd rename it to just `actions.fill('Edit')`

File: airgun/views/computeresource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abalakh) [reply airgun/views/computeresource.py]: No preference, both solutions are temporary until we find some time to learn conditional views.

File: airgun/widgets.py [hunk 6]
```diff
@@ -601,7 +601,6 @@ def read(self):
         """Returns a list of available actions."""
         return self.items
 
-
```
Comment (abalakh) [inline airgun/widgets.py]: That's a bad idea as it will cause pep8 checks to fail :) there should be 2 lines between classes, please revert this change.

**Comment by schlupov** [comment]
@abalakh @oshtaier  Please, take a look at last changes. Thank you.

File: airgun/entities/computeresource.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    def prerequisite(self, *args, **kwargs):
+        return self.navigate_to(self.obj, 'All')
+
+    def step(self, *args, **kwargs):
+        entity_name = kwargs.get('entity_name')
+        self.parent.search(entity_name)
+        self.parent.table.row(
+            name=entity_name)['Actions'].widget.fill('Edit')
+        self.parent.actions.fill('Edit')
```
Comment (abalakh) [inline airgun/entities/computeresource.py]: Seems like you forgot to remove this line from previous implementation, as there's no `actions` class attr anymore, it's part of table_column now.

**Comment by lhellebr** [comment]
Silva, don't you want to merge my changes first?

**Comment by schlupov** [comment]
@lhellebr changes merged

File: airgun/views/computeresource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (oshtaier) [reply airgun/views/computeresource.py]: Okay, so now we actually have implementation for conditional views. Can you look whether it can be helpful for you?

Examples:
https://github.com/SatelliteQE/airgun/blob/master/airgun/views/product.py#L150
https://github.com/SatelliteQE/airgun/blob/master/airgun/views/partitiontable.py#L48

**Comment by oshtaier** [comment]
so please check whether you can apply conditional views for different providers as they are really helpful for situation like this

File: airgun/views/computeresource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (schlupov) [reply airgun/views/computeresource.py]: I think it's not really necessary to use conditional views for every provider and I'd rather have it as it is. But If I really need to use conditional views what's the best approach? I don't know If I can use SatTab or anything else from views/common for every provider?

**Comment by oshtaier** [comment]
I looked on the application once more and yes it is strongly recommended to use conditional views in such situations. Actually it is not recommended, that is only one option you have
```
provider_type = FilteredDropdown(id='s2id_compute_resource_provider')
provider_content = ConditionalSwitchableView(reference='provider_type')

@provider_content.register('Docker')
class DockerProviderForm(View):
    description = TextInput(id='compute_resource_description')
    url = TextInput(id='compute_resource_url')
    user = TextInput(id='compute_resource_user')
    password = TextInput(id='compute_resource_password')
    email = TextInput(id='compute_resource_email')

@provider_content.register('EC2')
class EC2ProviderForm(View):
HTTP Proxy
Access Key
Secret Key
....

@provider_content.register('Google')
class GoogleProviderForm(View):
...
    
```

**Comment by schlupov** [comment]
@oshtaier  I edited view. I'm not sure how to change tests to use for example url for Docker provider cause 'DockerProviderForm.url': docker_url doesn't work in my test, it cannot locate field for url.

**Comment by oshtaier** [comment]
something like this:
```
provider_type: 'Docker'
provider_content: {'url': test, ...}
```
or 
```
provider_type: 'Docker'
provider_content.url: '',
```

**Comment by oshtaier** [comment]
That is an idea and power of conditional views, it will pass parameters to necessary view depends on what value is selected in list, you don't need to say explicitly `DOCKER` or `LIBVIRT`

**Comment by schlupov** [comment]
@oshtaier So I edited view and entities and in robottelo are my tests. Please check everything.

File: airgun/entities/computeresource.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    ResourceProviderDetailsView,
+)
+
+
+class ComputeResourceEntity(BaseEntity):
+
+    def create(self, values):
+        view = self.navigate_to(self, 'New')
+        view.fill(values)
+        if 'provider' in values.keys() and values['provider'] == 'oVirt' \
```
Comment (oshtaier) [inline airgun/entities/computeresource.py]: we are going to think about that block and whether conditional views can use `after_fill()` method

Comment (schlupov) [reply airgun/entities/computeresource.py]: Did you agree on something? Because now I'm just waiting for your opinion to move to issue #88. Should I change anything else here or in tests?

Comment (oshtaier) [reply airgun/entities/computeresource.py]: yes, I tested and solution works just perfect, let's me describe it for you. That should be last step I believe for that PR

Comment (oshtaier) [reply airgun/entities/computeresource.py]: Please, remove
```
if 'provider' in values.keys() and values['provider'] == 'oVirt' \
+                and 'provider_content.certification_authorities' not in \
+                values.keys():
+            view.provider_content.load_datacenters.click()
+        if 'provider' in values.keys() and values['provider'] == 'EC2':
+            view.provider_content.load_regions.click()
+        if 'provider' in values.keys() and values['provider'] == 'VMware':
+            view.provider_content.load_datacenters.click()
```
and then add corresponding block to each provider you need:
```
+    @provider_content.register('oVirt')
+    class oVirtProviderForm(View):
+        url = TextInput(id='compute_resource_url')
+        user = TextInput(id='compute_resource_user')
+        password = TextInput(id='compute_resource_password')
+        load_datacenters = Text("//*[contains(@id,'test_connection_button')]")
+        certification_authorities = TextInput(id='compute_resource_public_key')
     
          def after_fill(self, was_change):
              self.load_datacenters.click()
```
```
+    @provider_content.register('EC2')
+    class EC2ProviderForm(View):
+        http_proxy = TextInput(id='compute_resource_http_proxy_id')
+        access_key = TextInput(id='compute_resource_user')
+        secret_key = TextInput(id='compute_resource_password')
+        load_regions = Text("//*[contains(@id,'test_connection_button')]")

          def after_fill(self, was_change):
              self.load_regions.click()
```
```
+    @provider_content.register('VMware')
+    class VMwareProviderForm(View):
+        vcenter = TextInput(id='compute_resource_server')
+        user = TextInput(id='compute_resource_user')
+        password = TextInput(id='compute_resource_password')
+        display_type = Select(id='compute_resource_display_type')
+        vnc_console_passwords = Checkbox(
+            id='compute_resource_set_console_password')
+        enable_caching = Checkbox(id='compute_resource_caching_enabled')
+        load_datacenters = Text("//*[contains(@id,'test_connection_button')]")

          def after_fill(self, was_change):
              self.load_datacenters.click()
```

Comment (oshtaier) [reply airgun/entities/computeresource.py]: That shouldn't affect test logic as far as I see

**Comment by oshtaier** [comment]
And sorry, that issue wasn't good candidate for first issue =(

**Comment by oshtaier** [comment]
Once we finish - I propose to continue with that specific entity, not to move to another issues. You will have bunch tests to be ported, to get real value from current PR

File: airgun/views/computeresource.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    def is_displayed(self):
+        """Check if the right page is displayed"""
+        return self.browser.wait_for_element(
+            self.title, exception=False) is not None
+
+
+class ResourceProviderDetailsView(BaseLoggedInView):
+    name = TextInput(id='compute_resource_name')
+    description = TextInput(id='compute_resource_description')
+    api4 = Checkbox(id='compute_resource_use_v4')
```
Comment (oshtaier) [inline airgun/views/computeresource.py]: where did you find that checkbox?
image

Comment (schlupov) [reply airgun/views/computeresource.py]: @lhellebr This is your code. Where did you find it?

Comment (lhellebr) [reply airgun/views/computeresource.py]: That's a checkbox valid for RHEV provider in versions supporting RHEV APIv4.

Comment (schlupov) [reply airgun/views/computeresource.py]: So when you select provider RHV you can see checkbox Use APIv4 (experimental)

Comment (oshtaier) [reply airgun/views/computeresource.py]: so put that locator in corresponding provider, not on generic level

**Comment by schlupov** [comment]
@oshtaier I moved api4 and I changed views and entities. Everything ok now?

**Comment by oshtaier** [comment]
Is RHEV==oVirt?
Actually, I can't see oVirt at all in the list:
image

**Comment by schlupov** [comment]
Yes, RHEV=oVirt

**Comment by lhellebr** [comment]
RHEV is downstream of oVirt.

**Comment by oshtaier** [comment]
Okay, so your code will not work as there is no value like ovirt in the list, you should register views according to exact values

**Comment by abalakh** [comment]
@oshtaier nothing from my side, pending the last comment regarding registering view via proper values.

**Comment by schlupov** [comment]
@oshtaier I edited what you wanted. Please also check tests.

**Comment by oshtaier** [comment]
Looks good
Can you update robottelo PR with test results to make sure that your code here is working properly?

**Comment by schlupov** [comment]
I updated PR today in robottelo.

File: airgun/views/computeresource.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        load_zones = Text("//*[contains(@id,'test_connection_button')]")
+
+    @provider_content.register('Libvirt')
+    class LibvirtProviderForm(View):
+        url = TextInput(id='compute_resource_url')
+        display_type = Select(id='compute_resource_display_type')
+        console_passwords = Checkbox(
+            id='compute_resource_set_console_password')
+
+    @provider_content.register('OpenStack')
```
Comment (abalakh) [inline airgun/views/computeresource.py]: Shouldn't it be `RHEL OpenStack Platform`, according to @oshtaier's screenshot? I'm not trying to put yet another nitpick, but basically these register values aren't just some names like class names, they're used only if completely the same value was selected in dropdown, so if dropdown has something different - those extra fields won't be populated at all. Please make sure `register(...)` matches UI values, including case sensitivity.

**Comment by schlupov** [comment]
I changed that in views. It's the only place in my code where I use it because open stack can't be really tested insomuch as in settings is no option for open stack.
